### ISSUE TITLE
code replicates error in vendor documentation

### ISSUE THREAD

**Issue by pabigot** [sparkfun/Qwiic_BME280_CCS811_Combo#8]
Around [here](https://github.com/sparkfun/Qwiic_BME280_CCS811_Combo/blob/3e43b3289bce05ccb271034105f0ce0e1a6ca11d/Libraries/CCS811/src/SparkFunCCS811.cpp#L467) the code for configuring the CCS811 ENV_DATA values faithfully represents the instructions in the [CCS811 Programming guide](https://ams.com/documents/20143/36005/CCS811_AN000369_2-00.pdf/25d0db9a-92b9-fa7f-362c-a7a4d1e292be) section 16, being:
```
i2c_buff[0] = ((RH % 1000) / 100) > 7 ? (RH/1000 + 1)<<1 : (RH/1000)<<1;
i2c_buff[1] = 0;
if(((RH % 1000) / 100) > 2 && (((RH % 1000) / 100) < 8)) {
  i2c_buff[0] |= 1;
}
```
The demonstration code is wrong, as it results in storing a value that is one unit too large when the fractional part is strictly between 0.7 and 0.8.  In that case the whole value is incremented and the half-value is added, applying the half-round operation twice.

I don't use this library, but found this while trying to figure out why writing `ENV_DATA` doesn't work on my board (the I2C read of address 0x05 sets `READ_REG_INVALID` in `ERROR_ID`), and figured somebody might find this useful.

**Comment by nseidle** [comment]
I think I get what you're saying.

Let's run 42.745 through the code. 

    envData[0] = ((rH % 1000) / 100) > 7 ? (rH / 1000 + 1) << 1 : (rH / 1000) << 1;

This puts (42<<1) in envData[0]. Then 

    if (((rH % 1000) / 100) > 2 && (((rH % 1000) / 100) < 8))
        envData[0] |= 1; //Set 9th bit of fractional to indicate 0.5%

This puts the 0.5 into bit 0. So the outcome is 42.5%. Looks good.

Let's run 42.856 through.

    envData[0] = ((rH % 1000) / 100) > 7 ? (rH / 1000 + 1) << 1 : (rH / 1000) << 1;

This puts (43<<1) in envData[0]. Then 

    if (((rH % 1000) / 100) > 2 && (((rH % 1000) / 100) < 8))
        envData[0] |= 1; //Set 9th bit of fractional to indicate 0.5%

This puts the 0.5 into bit 0. So the outcome is 43.5%. That's too much.

So I think the solution is:

    envData[1] = 0; //CCS811 only supports increments of 0.5 so bits 7-0 will always be zero

    //Split value into 7-bit integer and 9-bit fractional
    if( (rH % 1000) / 100 > 7)
    {
      envData[0] = (rH / 1000 + 1) << 1; //Increase 42.756 to 43.0
    }
    else
    {
      envData[0] = (rH / 1000) << 1; //Store 42
      if ( (rH % 1000) / 100 > 2)
        envData[0] |= 1; //Set 9th bit of fractional to indicate 0.5% - 42.5%
    }

**Comment by pabigot** [comment]
Yeah, you could do something like that.  Or just rely on integer truncation and do:
```
envData[0] = (rH + 250) / 500;
```
which rounds to the nearest half unit.

**Comment by nseidle** [comment]
Much sneakier. Thanks - updating lib.

**Comment by pabigot** [comment]
You're welcome.

If anybody else gets here because attempting to read the `ENV_DATA` register fails, my conclusion is that the register is not readable: it fails with both app firmware 1.1.0 (which was on the Sparkfun board I got long ago), and with 2.0.0.  Which is pretty annoying, since I'd like to know whether the power-on state has sane defaults (see NordicSemiconductor/Nordic-Thingy52-FW#21 for background).

Also, initial experience with 2.0.0 appears to fix a variety of issues with the status, and maybe baseline maintenance.

**Comment by hmhalvorsen** [comment]
> Let's run 42.856 through.
> 
> ```
> envData[0] = ((rH % 1000) / 100) > 7 ? (rH / 1000 + 1) << 1 : (rH / 1000) << 1;
> ```
> This puts (43<<1) in envData[0]. Then
> 
> ```
> if (((rH % 1000) / 100) > 2 && (((rH % 1000) / 100) < 8))
>     envData[0] |= 1; //Set 9th bit of fractional to indicate 0.5%
> ```
> This puts the 0.5 into bit 0. So the outcome is 43.5%. That's too much.

I am not sure if this is correct. ((42856 mod 1000) / 100) = 8.56, which is not less than 8. Surely we wouldn't set the 9th bit here?

**Comment by hmhalvorsen** [comment]
Yes, that makes sense. Thank you for clarifying this matter.

**Comment by fbraswell** [comment]
Why to go all the trouble of the code in the data sheet, which tries to keep the 8 bits of fraction zero? The direct way to populate the two %RH registers is to multiply the RH value by 512, convert the value to an int and split it in two. Here's the code in the Squirrel language for the Electric Imp processor:

        // The 16 bit %RH register is composed of a 7 bit integer portion and 9 bit 
        // fractional portion. Multiplying the %RH by 512 effectively populates the
        // integer and fractional portion.
        local h = hval * 512;
        // Next, split the two bytes of the integer apart for sending to the ENV_DATA registers
        local h_hi = h.tointeger() >> 8; // get upper 8 bits
        local h_lo = h.tointeger() & 0x00FF; // grab lower 8 bits

The temperature calculation adds the 25 degree offset:

        // First, convert the tempF to tempC
        local tvalC = (tvalF - 32.0) * 5.0 / 9.0;
        // In this case, there is an offset of 25 deg C. This means that if the 2 registers are 0
        // then the temperature is -25 deg C
        // For the computation, add the offset of 25 and multiply by 512, as we did with the %RH above
        local t = (tvalC + 25.0) * 512;
        // Split the integer into two parts for sending the the ENV_DATA registers
        local t_hi = t.tointeger() >> 8; // get upper 8 bits
        local t_lo = h.tointeger() & 0x00FF; // grab lower 8 bits

**Comment by 532910** [comment]
[CCS811 datasheet](https://ams.com/documents/20143/36005/CCS811_DS000459_7-00.pdf/3cfdaea5-b602-fe28-1a14-18776b61a35a) says:
> Temperature / Humidity is stored as an unsigned 16 bits in 1/512



Could you explain, why

> bits 7-0 will always be zero

?
